src/coreclr/jit/codegenriscv64.cpp

```diff
@@ -7143,11 +7457,164 @@ void CodeGen::genEstablishFramePointer(int delta, bool reportUnwindData)
 }
 
 //------------------------------------------------------------------------
-// genAllocLclFrame: Probe the stack and allocate the local stack frame: subtract from SP.
+// genStackProbe: Probe the stack without changing it
+//
+// Notes:
+//      This function is using loop to probe each memory page.
+//
+// Arguments:
+//    frameSize - total frame size
+//    rOffset - usually initial register number
+//    rLimit - an extra register for comparison
+//    rPageSize - register for storing page size
+//
+void CodeGen::genStackProbe(ssize_t frameSize, regNumber rOffset, regNumber rLimit, regNumber rPageSize)
+{
+    // make sure frameSize safely fits within 4 bytes
+    noway_assert((ssize_t)(int)frameSize == (ssize_t)frameSize);
+
+    // According to RISC-V Privileged ISA page size should be equal 4KiB
+    noway_assert(compiler->eeGetPageSize() == 0x1000);
+
+    GetEmitter()->emitLoadImmediate(EA_PTRSIZE, rLimit, -frameSize);
+    regSet.verifyRegUsed(rLimit);
+
+    GetEmitter()->emitIns_R_I(INS_lui, EA_PTRSIZE, rPageSize, compiler->eeGetPageSize() >> 12);
+    regSet.verifyRegUsed(rPageSize);
+
+    GetEmitter()->emitIns_R_R_I(INS_addi, EA_PTRSIZE, rOffset, REG_SPBASE, 0);
+
+    // Loop:
+    GetEmitter()->emitIns_R_R_R(INS_sub, EA_PTRSIZE, rOffset, rOffset, rPageSize);
+    // tickle the page - Read from the updated SP - this triggers a page fault when on the guard page
+    GetEmitter()->emitIns_R_R_I(INS_lw, EA_4BYTE, REG_R0, rOffset, 0);
 
+    // each instr is 4 bytes
+    // if (rOffset >= rLimit) goto Loop;
+    GetEmitter()->emitIns_R_R_I(INS_bge, EA_PTRSIZE, rOffset, rLimit, -2 << 2);
```

Comment:
IMO, `rLimit` should be  `-frameSize + REG_SPBASE` and initial value of rOffset should be `REG_SPBASE - rPageSize`? Could you please check?

Reply:
> IMO, `rLimit` should be `-frameSize + REG_SPBASE` and initial value of `rOffset` should be `REG_SPBASE - rPageSize`

Nice catch! Thanks, of course it should be `-frameSize + REG_SPBASE`. After the update, the assembly should look more or less like this:

```asm
; load   rLimit, -frameSize
add  rLimit, rLimit, REG_SPBASE
; rLimit <-- (-frameSize + REG_SPBASE)
lui  rPageSize, pageSize >> 12
; rPageSize <-- 0x1000

sub  rOffset, REG_SPBASE, rPageSize
; rOffset <-- (REG_SPBASE - rPageSize)

; loop:
  lw   R0, rOffset ; tickle the page
  sub  rOffset, rOffset, rPageSize
  bge  rOffset, rLimit, -2 << 2

```